src/main/js/utils/generate.js

```diff
@@ -6,57 +6,109 @@ import {
+            j -= 1;
+          }
+          schedules[j + 1] = key;
+        }
+      }
+    } catch (e) {
+      schedules = e;
+    }
+  }
+  return schedules;
+};
+
+const fetchShortUrl = (courseList, formValues) => {
+  const queryObject = {
+    ...formValues,
+    c: courseList.map((course) => {
+      const name = `${course[0].subject}${course[0].courseNumber}`;
+      return JSON.stringify({
+        name,
+        ...(formValues.courses ? formValues.courses[`${name}${course.id}`] : {}),
+      });
+    }),
+  };
+  delete queryObject.courses;
+  return fetch(`${shortenUrl}?${stringify({ data: JSON.stringify(queryObject), prefix: 'generator' })}`)
+    .then(response => response.json());
+};
+
+const redirectToGenerator = (schedules, search = '') => startRedirect({
+  pathname: '/generator',
+  hash: schedules.length > 0 ? '#schedules' : '#timetable',
+  search: search.replace('==', '='),
+});
+
+const generateSchedules = (values, loadQuery = '') => (dispatch, getState) => {
   const { courseList } = getState().courses;
   if (courseList.length !== 0) {
     dispatch(startGenerating());
+
+    const shortCode = loadQuery || !values.genURL ? Promise.resolve(loadQuery.substring(2))
+      : fetchShortUrl(courseList, values);
+
     const filteredResults = filterCourses(values, courseList);
     dispatch(filteredCourses(filteredResults));
-    const filteredCourseList = filteredResults.map(result => (
-      result.filtered));
-    let schedules = [];
-    if (!filteredCourseList.find(list => list.length === 0)) {
-      const scheduleMaker = new ScheduleMaker(filteredCourseList, parseInt(values.gap, 10));
-      try {
-        schedules = scheduleMaker.makeSchedules();
-      } catch (e) {
-        schedules = e;
-      }
-    }
-    const queryObject = {
-      ...values,
-      c: courseList.map((course) => {
-        const name = `${course[0].subject}${course[0].courseNumber}`;
-        return JSON.stringify({
-          name,
-          ...(values.courses ? values.courses[`${name}${course.id}`] : {}),
-        });
-      }),
-    };
-    delete queryObject.courses;
-    (loadQuery ? Promise.resolve() : fetch(`${shortenUrl}?${stringify({ data: JSON.stringify(queryObject), prefix: 'generator' })}`))
-      .then(response => (loadQuery ? loadQuery.substring(2) : response.json()))
-      .then((result) => {
-        const querystring = `q=${result}`;
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: result ? querystring : '',
-        }));
-      })
-      .catch(() => {
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: '',
+    let filteredCourseList = filteredResults.map(result => result.filtered);
+
+    const gpasLoading = values.sortByGPA === 'yes' ? addGPA(filteredCourseList) : Promise.reject();
```

Comment:
Put this call before filtering, actually before `shortCode`. Does it make sense why we should do this?

Reply:
Fixed.